### PR TITLE
ItemsList: supply item actions from view model

### PR DESCRIPTION
**Pull request by jacobsimeon**
This lays the ground work so that `ArchivedItemsListViewModel` can supply different actions from what `MyListItemsListViewModel` uses.

One big question is around the re-use of `ReadableAction`. I went ahead and adopted it for this use case but I'm also wondering if 1) a separate struct should be use to model the actions or 2) if we can simply supply `UIAction` objects directly.

### CODE REVIEW

File: PocketKit/Sources/PocketKit/MyList/ItemsList/ItemsListViewModel.swift [hunk 1]
```diff
@@ -35,10 +35,11 @@ protocol ItemsListViewModel: AnyObject {
...
     func selectCell(with: ItemsListCell<ItemIdentifier>)
-    func shareItem(with: ItemsListCell<ItemIdentifier>)
 
-    func toggleFavorite(_ cell: ItemsListCell<ItemIdentifier>)
-    func archive(_ cell: ItemsListCell<ItemIdentifier>)
-    func delete(_ cell: ItemsListCell<ItemIdentifier>)
+    func shareAction(for objectID: ItemIdentifier) -> ReadableAction?
+    func favoriteAction(for objectID: ItemIdentifier) -> ReadableAction?
+    func overflowActions(for objectID: ItemIdentifier) -> [ReadableAction]?
+    func swipeActions(for objectID: ItemIdentifier) -> [UIContextualAction]
```
Comment (dskuza) [inline PocketKit/Sources/PocketKit/MyList/ItemsList/ItemsListViewModel.swift:42]:
```suggestion
    func trailingSwipeActions(for objectID: ItemIdentifier) -> [UIContextualAction]
```

A small detail, but at least in our current usage, the swipe actions we provide are for the trailing side of the cell. Being a little more explicit in the API might just help in clarity, especially if we may want to add leading actions one day.

File: PocketKit/Sources/PocketKit/MyList/ItemsList/ItemsListItemCell.swift [hunk 2]
```diff
@@ -241,11 +202,24 @@ extension ItemsListItemCell {
...
-        favoriteButton.configuration?.image = state.model?.favoriteButtonImage
-        favoriteButton.accessibilityLabel = state.model?.favoriteButtonAccessibilityLabel
 
-        thumbnailView.image = nil
+        favoriteButton.accessibilityLabel = state.model?.favoriteAction?.title
+        favoriteButton.accessibilityIdentifier = state.model?.favoriteAction?.accessibilityIdentifier
+        favoriteButton.configuration?.image = state.model?.favoriteAction?.image
+
+        if let favoriteAction = UIAction(state.model?.favoriteAction) {
+            favoriteButton.addAction(favoriteAction, for: .primaryActionTriggered)
```
Comment (dskuza) [inline PocketKit/Sources/PocketKit/MyList/ItemsList/ItemsListItemCell.swift:211]: Do we need to remove all actions before adding an action? I just want to verify that we won't be stacking actions for the same control event.

**Review by dskuza — COMMENTED** [review]
Overall, I like these changes. 

I don't mind having these models return `ReadableAction` types, primarily because it keeps `UIKit` as an implementation detail of the view (controller) using it. It also allows us to do things like this:

```swift
favoriteButton.accessibilityLabel = state.model?.favoriteAction?.title
favoriteButton.accessibilityIdentifier = state.model?.favoriteAction?.accessibilityIdentifier
```

and keep these definitions within a single type, where `UIAction`'s API doesn't expose these sorts of details.

The above is considering that I'm understanding the PR's description correctly, and the question is whether to have the _models_ return these types. 

Added comment: if we are reusing these actions, maybe we need to have a conversation about the `Readable` name, and where / whether it makes sense to use.

Happy to go over this with you live, as always!

File: PocketKit/Sources/PocketKit/MyList/ItemsList/ItemsListItemCell.swift (hunk 2 above)
Comment (jacobsimeon) [reply PocketKit/Sources/PocketKit/MyList/ItemsList/ItemsListItemCell.swift:211]: From Apple's docs:

> UIActions are uniqued based on their identifier, and subsequent actions with the same identifier replace previously added actions.

So I take that to mean that we are safe here. This is why the `.favorite()` and `.unfavorite()` helpers use the same identifier.

**Comment by jacobsimeon** [comment]
>  maybe we need to have a conversation about the Readable name

Yeah I agree. Ideas for names? I was thinking maybe `ItemAction`

**Comment by dskuza** [comment]
`ItemAction` works great - that's what I was thinking, as well.

File: PocketKit/Sources/PocketKit/ItemAction.swift [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            image: UIImage(asset: .favoriteFilled)
+                .withTintColor(UIColor(.branding.amber4), renderingMode: .alwaysOriginal),
+            handler: handler
+        )
+    }
+
+    static func share(_ handler: @escaping () -> ()) -> ItemAction {
+        return ItemAction(
+            title: "Share",
+            // intentionally the same as `favorite()` since we want to replace
```
Comment (dskuza) [inline PocketKit/Sources/PocketKit/ItemAction.swift:86]: Perhaps this is a byproduct of a copy/paste for (un)favorite actions.

**Review by dskuza — APPROVED** [review]

File: PocketKit/Sources/PocketKit/ItemAction.swift (hunk 3 above)
Comment (jacobsimeon) [reply PocketKit/Sources/PocketKit/ItemAction.swift:86]: yup

**Pull request merged**